metricbeat/module/prometheus/collector/collector.go

```diff
@@ -140,3 +160,13 @@ func (m *MetricSet) addUpEvent(eventList map[string]common.MapStr, up int) {
 	}
 
 }
+
+func matchMetricFamily(family string, matchMetrics *[]string) bool {
+	for _, checkMetric := range *matchMetrics {
+		matched, _ := regexp.MatchString(checkMetric, family)
```

Comment:
We should pre-parse patterns when creating the metricset, so we earlier verify that they are valid, and we don't have to parse them for every metric.

Reply:
hmm, could you elaborate more on this please? What kind of validation you have in mind? We will still need to loop over the `matchMetrics` for every `metricFamily` so as to check if it matches any of the given filters, no?

Reply:
`regexp.MatchString(pattern, family)` compiles the pattern on every call, this has two problems:
* The pattern can be invalid and we would only detect it here, but it would be a configuration problem that should have been detected when instantiating the metricset.
* Compiling all patterns for each metric can be expensive.

I propose to pre-compile with `regexp.Compile()` all the patterns already on the metricset constructor (on `New`), and store the compiled `regexp.Regexp` objects instead of the string patterns.  This will serve the double purpouse of earlier checking that they are valid, and to have them precalculated.

Reply:
Match metrics would be passed to this method as `[]*regexp.Regexp`, and we would do here something like this:
```suggestion
		matched := checkMetric.MatchString(family)
```

Reply:
Legit! Thanks! 🙂